fetch.bs

```diff
@@ -2353,7 +2435,374 @@ X-Content-Type-Options           = "nosniff" ; case-insensitive</pre>
+ <code>TokenBinding.tokenbinding_id</code> field be set to
+ <var>tokenBindingId</var>.
+ <li>Let <var>tokenBinding</var>'s <code>TokenBinding.signature</code> be set
+ to the result of computing a signature over <var>tokenBindingType</var>,
+ <var>tokenBindingId</var>'s <code>key_parameters</code> field and
+ <var>tokenBindingEKM</var> using <var>tokenBindingKeyPair</var> as described in
+ <a href="https://tools.ietf.org/html/draft-ietf-tokbind-protocol#section-3.3">section 3.3</a>
+of [[!TOKBIND-PROTOCOL]].
+ <li>Return <var>tokenBinding</var>.
+</ol>
+
+<h4 id=computing-token-binding-header-value>Computing a Token Binding Header Value</h4>
+
+<p><dfn abstract-op export id=concept-compute-token-binding-header-value>Compute a Token Binding header value</dfn>
+that can be passed in a `<a http-header><code>Sec-Token-Binding</code></a>`
+<a for=/>header</a> for a <a for=/>request</a>
+<var>httpRequest</var> and a <a>connection</a> <var>tlsConnection</var>,
+using the <a for=/>token-binding key store</a> and the
+<a>connection pool</a>, by running these substeps:
+
+<ol>
+ <li><p>Return null if any of the following are true.
+
+ <ol>
+  <li><p><var>httpRequest</var>'s <a for=request>use-token-binding flag</a> is not set.
+
+  <li><p><a for=connection>token-binding protocol version</a> in <var>tlsConnection</var> is null.
+ </ol>
+
+ <li><p>Let <var>tokenBindingMessage</var> be a <code>TokenBindingMessage</code> struct from
+<a href="https://tools.ietf.org/html/draft-ietf-tokbind-protocol#section-3">section 3</a>
+ of [[!TOKBIND-PROTOCOL]].
+
+ <li><p>Let <var>providedTokenBindingKeyParameters</var> be
+ <var>tlsConnection</var>'s <a for=connection>token-binding key parameters</a>.
+
+ <li><p>Let <var>providedTokenBindingKeyPair</var> be the result of
+<a abstract-op lt="Get the token-binding key">getting the token-binding key</a>
+for the <a for=/>origin</a> of <var>httpRequest</var>'s <a for=request>current url</a>
+and <var>providedTokenBindingKeyParameters</var>.
+
+ <li><p>Let <var>providedTokenBindingId</var> be the result of
+ <a abstract-op lt="Compute a Token Binding ID">computing a token binding ID</a>
+from <var>providedTokenBindingKeyPair</var> and
+<var>providedTokenBindingKeyParameters</var>.
+
+ <li><p>Let <var>providedTokenBinding</var> be the result of
+<a abstract-op lt="Compute a Token Binding">computing a token binding</a>
+using <a for=/>token-binding type</a> <code>provided_token_binding</code>,
+<var>providedTokenBindingId</var>, <var>providedTokenBindingKeyPair</var>,
+and <var>tlsConnection</var>'s
+<a for=connection>token-binding Exported Keying Material</a>.
+
+ <li><p>Append <var>providedTokenBinding</var> to
+ <var>tokenBindingMessage</var>'s <code>TokenBindingMessage.tokenbindings</code> field.
+
+ <li><p>If <var>httpRequest</var>'s <a for=request>referred-token-binding origin</a> is not null,
+ then run these substeps:
+  <ol>
+   <li><p>Let <var>referredTlsConnection</var> be the result of
```

Comment:
Continuing a comment thread from the previous PR:

I still don't like that referredTlsConnection is created, when all it's needed for is the key type. For the case of a referred Token Binding via an HTTP header, I think the most sensible thing to do is use the same key type/key as was used for the request that received the HTTP response header. This is behavior that doesn't require opening a new connection.

Supporting the fetch() API's mechanism for referred Token Bindings might have some rule that's this simple, like "use the last used key/key type for that origin", but I don't know if that's the rule we'd actually want to write down for this spec. I think we can write a much simpler "Computing a Token Binding Header Value" that doesn't rely on creating this referredTlsConnection if we limit this PR to exclude changes to the fetch() API.

(On the previous PR, the rationale for why the referredTlsConnection is needed was that the server might change the key type (key parameters) it wants to use. Servers changing the Token Binding key parameters they support is fraught with peril and most likely broken if the server wants to have any chance of keeping tokens bound to some Token Binding key.)

Reply:
I removed fetch() API related changes.

For referred Token Binding using redirects, the referredTlsConnection will already exist - it will just be looked up.